portmapper.go

```diff
@@ -93,49 +86,79 @@ func Unregister(name string, port int) error {
 	return nil
 }
 
-// Loops over each service type in the map and attempt to register
-func RegisterServices() {
-	for {
-		wg := &sync.WaitGroup{}
-		for service_name := range ServiceMap {
-			wg.Add(1)
-			go func() {
-				defer wg.Done()
-				println("registering: " + service_name)
-				etcdClient = etcd.NewClient([]string{EtcdHost})
-				svc := ServiceMap[service_name].Service
-				ServiceMap[service_name].Timestamp = time.Now().Unix()
-				ServiceMap[service_name].err = nil
-
-				if err := svc.validate(); err != nil {
-					ServiceMap[service_name].err = err
-				}
-
-				bytes, err := svc.Marshal()
-				if err != nil {
-					ServiceMap[service_name].err = err
-				}
-
-				if _, err = etcdClient.Set(svc.path(), string(bytes), 0); err != nil {
-					ServiceMap[service_name].err = err
-				}
-			}()
-		}
-
-		// Wait for all services to be registered
-		wg.Wait()
-		time.Sleep(60 * time.Second)
-	}
+// legacy register
+// Register a (service, port) tuple.
+func Register(name string, port int) error {
+	return RegisterService(name, port, 11) // about 4 seconds
 }
 
 // Register a (service, port) tuple.
-func Register(name string, port int) error {
+func RegisterService(name string, port int, retries int) error {
 	svc := &Service{name, port, os.Getenv("HOSTNAME")}
-	ServiceMap[name] = &ServiceRegistration{Service: svc, Timestamp: 0, err: nil}
 
-	//XXX I assume that this will not exit ever
-	once.Do(RegisterServices)
-	return nil
+	var regerr error
+
+	for try := 1; try < retries; try++ {
```

Comment:
You should differentiate between transient and permanent failures. E.g. if the service fails validation, fail fast without retrying. Same with the Marshal() call. 

The only thing that should retry is the interaction with Etcd.